### DISCUSSION TITLE
Uncaught (in promise) TypeError: Failed to fetch

### DISCUSSION THREAD

**Discussion by yufeng1524** [gildas-lormeau/zip.js#432] [Q&A]
Hi

we are using zipjs 2.6.63, and meet below error when zip file for sometimes, but not every time:

**Uncaught (in promise) TypeError: Failed to fetch**

webworker is configured for zipjs.

Additionally,  we found there's same issue in #370 , and the issue has been fixed in 2.6.37 which is older than we used. And we didn't abort zip writer when meet the error.

Any idea about this issue?

**Comment by gildas-lormeau** [comment]
Why don't you use the last version?

**Comment by yufeng1524** [comment]
> Why don't you use the last version?

thanks for quick responding.

by the way, is it a known issue?

**Comment by gildas-lormeau** [comment]
This issue should be fixed but it's a bit old. I guess the version 2.6.82 fixes your issue, see https://github.com/gildas-lormeau/zip.js/releases/tag/v2.6.82. It's very similar to the fix in the version 2.6.37, see https://github.com/gildas-lormeau/zip.js/commit/9ce18f61902de331245e6f5fbbadf4ae75b67683 (v2.6.82) and https://github.com/gildas-lormeau/zip.js/commit/e8e5269541f2eb3a58e4ec3c23c8ef91fa32cff9 (v2.6.37).

**Reply by peaceps** [reply]
Hi gildas-lormeau, I also meet this 'Failed to fetch' error once when trying to zip files on chrome with latest 2.7.17 without bufferedWrite option. 
And after this error occurs I won't get the zipped content any more.

**Reply by tannerolsen7** [reply]
I am also getting this error when trying to zip files on chrome with the latest 2.7.17 version.

**Reply by gildas-lormeau** [reply]
I cannot help you without any additional information. For example, you could post the stack trace. All the tests in zip.js run fine in Chrome without triggering any exception.

**Reply by tannerolsen7** [reply]
Hi @gildas-lormeau, I apologize for the cryptic message. I haven't reached out in a discussion like this before. Does the below stack trace help? I am going through an array of photos and adding them and it seems to be triggered by the number of photos that are added. I haven't found the exact cause or a number of photos that causes this yet. 

Uncaught (in promise) TypeError: Failed to fetch
    at sendFetchRequest (io.js:415:25)
    at getContentLength (io.js:404:26)
    at initHttpReader (io.js:344:9)
    at FetchReader.init (io.js:269:9)
    at HttpReader.init (io.js:474:21)
    at initStream (io.js:663:16)
    at addFile (zip-writer.js:262:9)
    at ZipWriter.add (zip-writer.js:144:21)
    at createZip (generateFiles.ts:93:27)

**Reply by gildas-lormeau** [reply]
Thank you @tannerolsen7. The error is the one triggered by the native `fetch` function actually (see below). Do you pass any option when creating the zip file or adding entries in the zip file?

```
...
at sendFetchRequest (io.js:415:25)
...
```

https://github.com/gildas-lormeau/zip.js/blob/b519130fd7d4c3bea84122ecb1230864b5cd18e4/lib/core/io.js#L415

**Reply by tannerolsen7** [reply]
Here is a condensed version of my code. I am testing functionality and have a password hard-coded for now:

```
    const zipFileWriter = new BlobWriter();
    const zipWriter = new ZipWriter(zipFileWriter, {password: "test123"});

    for(let exam of data) {
        // exam PDF
        
        //exam images
          for(let i = 0; i < imgs.length; i++){
              const imgReader = new HttpReader(imgs[i])
              zipWriter.add(`${id}/${id}-${i}.jpg`, imgReader)
          }
    const csvReader = new TextReader(summaryCsv);
    await zipWriter.add("summary.csv", csvReader);
    await zipWriter.close();

    // Retrieves the Blob object containing the zip content into `zipFileBlob`.
    const zipFileBlob = await zipFileWriter.getData();
```

The image is coming from a url. Could that be my issue?

**Reply by gildas-lormeau** [reply]
Could you try to replace the `for/let` loop on `imgs` with a `for/of` loop (or a `Promise.all`/`map` loop to add entries in parallel), and add the missing `await` keyword before `zipWriter.add(...)` in the loop?

**Reply by tannerolsen7** [reply]
Yes, that makes sense. I'll try that and get back to you!

**Reply by yufeng1524** [reply]
Hi gildas-lormeau，

After we upgrade zipjs to latest version, the issue happened again. We found that the issue can be reproduced easily when there is little hardware disk space  left.  For example,  we need zip files about 1G (contains much more files in the zip, we add entries one by one), if there's 3G disk space left, the issue would happen many times.

Is there some requirements about the minimum disk space when using zipjs?

**Reply by gildas-lormeau** [reply]
No, there's no real limitation, it's something that's managed by the environment. zip.js doesn't even know that a disk exists actually.

**Reply by tannerolsen7** [reply]
@gildas-lormeau I am still running into the same issues with the above suggestions. 

```
        Promise.all(
            imgs.map(async (url) => {
                const imgReader = new HttpReader(url)
                await zipWriter.add(`${id}/${url}.jpg`, imgReader)
            })
        )
```
OR
```
            for (let url of imgs) {
                const imgReader = new HttpReader(url)
                await zipWriter.add(`${id}/${url}.jpg`, imgReader)
            }
```
To ensure it's not an issue from my end, is this what you meant?

**Reply by gildas-lormeau** [reply]
Indeed, I confirm this is what I meant. You could maybe optimize RAM usage by implementing your own [`Writer`](https://github.com/gildas-lormeau/zip.js/blob/a64c5238c6bd20b64b6aa32551e80347289ac5fb/lib/core/io.js#L90-L110) class or by relying on the [`WritableStream`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream) API instead of using `BlobWriter` (for `zipFileBlob`). To help you, I would need to know what you do with `zipFileBlob` (e.g. save on disk, upload on network). Also, what browser(s)/JS engine(s) must you support?

**Reply by yufeng1524** [reply]
Here is part of code, we add files in one zip one by one, and  meet the exception : Uncaught (in promise) TypeError: Failed to fetch
could you help to take a look?

const zipJS = require('@zip.js/zip.js');
zipJS.configure({useWebWorkers: true});

writer = new zipJS.ZipWriter(new zipJS.BlobWriter());

await writer.add(fileName, new BufferReader(file))

class BufferReader extends zipJS.Uint8ArrayReader {
    constructor(buffer) {
        super(new Uint8Array(buffer));
    }
}

**Reply by gildas-lormeau** [reply]
I'm sorry but I cannot help you. You did not answer to my question and the code does not give me any additional clue on what you do with `zipFileBlob`.

**Comment by gildas-lormeau** [comment]
I'm closing the Discussion because it should be an issue. If you create an issue, please provide detailed info about the issue.
